services/web/server/src/simcore_service_webserver/projects/projects_db.py

```diff
@@ -819,10 +656,73 @@ async def check_delete_project_permission(self, user_id: int, project_uuid: str)
                     conn, user_id, project_uuid, for_update=True
                 )
                 # if we have delete access we delete the project
-                user_groups: list[RowProxy] = await self.__load_user_groups(
+                user_groups: list[RowProxy] = await self._list_user_groups(
                     conn, user_id
                 )
-                _check_project_permissions(project, user_id, user_groups, "delete")
+                check_project_permissions(project, user_id, user_groups, "delete")
+
+    #
+    # Project TAGS
+    #
+
+    async def add_tag(self, user_id: int, project_uuid: str, tag_id: int) -> dict:
+        async with self.engine.acquire() as conn:
+            project = await self._get_project(conn, user_id, project_uuid)
+            # pylint: disable=no-value-for-parameter
+            query = study_tags.insert().values(study_id=project["id"], tag_id=tag_id)
+            user_email = await self._get_user_email(conn, user_id)
+            async with conn.execute(query) as result:
+                if result.rowcount == 1:
+                    project["tags"].append(tag_id)
```

Comment:
what happens if the tag already exists?

Reply:
yes, this old piece of code is not very reliable.. I will add a bit of love to it ;-)

Reply:
... but not too much. I will refactor more later in https://github.com/ITISFoundation/osparc-simcore/issues/3516